engine/backends/coq/coq_backend.ml

```diff
@@ -8,6 +8,8 @@ module CoqBackend = struct
   module InputLanguage = struct
     open Features
     include Off
+    (* include On.Monadic_action *)
+    include On.Early_exit
```

Comment:
Are you translating early exits for real, or is that just for debugging purposes?

Reply:
I wanted to start trying, however it seems there are still things to do, before we can actually try translating to the backends?

Reply:
Well, yeah, I'm currently writing a phase that transforms `return`s, `continue`s and `break`s into monadic let bindings.
I think that should really be done in phases, not in backends (backends should be as simple as possible)

Actually, I would like to have multiple phases for addressing this:
 1. one fully monadic
 2. one that just rewrite the control flow, basically that tries to remove early `return`s by pushing them at the end of the function (this phase might fail, sometimes)
 3. one that choose between those two

I'm currently thinking about (1), but if you're interested in writing phases, maybe you could help me with (2) :)
If you're interested we can chat about that tomorrow at the meeting

Reply:
I just wanted to start writing the translation into monadic binds, as it currently only works for result and option types, however V2 allows more general early returns, which I could handle directly as a state monad in Coq.

Reply:
Oh right! Then you might want to disable `early_exit` but enable `monadic_binding`.
That allows for expr `Let` to be monadic.
(But fow now, it's draft and no phase actually spit out `monadic_binding` nodes)

Reply:
oh, you mean that you want to translate directly early returns into monadic stuff in Coq directly in the backend?
I think this is a nontrivial thing that should be a phase (and I hope the phase I'm writing corresponds to what you describe)

Reply:
Nice! That's what I plan to output! (with explicit binds though)

Reply:
The complicated part is to make breaks, continues, returns, local mutation and functionalization of loops all play nicely together I think

Reply:
I have done that already for ssprove, so that is why I could try and play with it directly ..

Reply:
The interesting part of the implementation can be found here:
https://github.com/cmester0/hacspec/blob/7c6d21cd45ec2625d8dfe32505388cde6c5d496c/coq_ssprove/src/Hacspec_Lib.v#L1458-L1553
https://github.com/cmester0/hacspec/blob/7c6d21cd45ec2625d8dfe32505388cde6c5d496c/coq_ssprove/src/Hacspec_Lib.v#L1718-L1726